From 46006bde3c7f76e2200112ec876d016cac2ac2c1 Mon Sep 17 00:00:00 2001
From: Thierry FOURNIER <tfournier@exceliance.fr>
Date: Fri, 21 Mar 2014 21:45:15 +0100
Subject: [PATCH 111/111] MINOR: pattern: Add function to prune and reload
 pattern list.

This function it is used for dynamically update all the patterns
attached to one file. This function is atomic. All parsing or indexation
failures are reported in the haproxy logs.
---
 include/proto/pattern.h |  2 ++
 src/pattern.c           | 29 +++++++++++++++++++++++++++++
 2 files changed, 31 insertions(+)

diff --git a/include/proto/pattern.h b/include/proto/pattern.h
index 57294d8..2e1c918 100644
--- a/include/proto/pattern.h
+++ b/include/proto/pattern.h
@@ -180,6 +180,8 @@ int pat_ref_delete(struct pat_ref *ref, const char *key);
 int pat_ref_delete_by_id(struct pat_ref *ref, struct pat_ref_elt *refelt);
 void pat_ref_prune(struct pat_ref *ref);
 int pat_ref_load(struct pat_ref *ref, struct pattern_expr *expr, int patflags, int soe, char **err);
+void pat_ref_reload(struct pat_ref *ref, struct pat_ref *replace);
+
 
 /*
  * pattern_head manipulation.
diff --git a/src/pattern.c b/src/pattern.c
index 16fb704..3238fbb 100644
--- a/src/pattern.c
+++ b/src/pattern.c
@@ -19,6 +19,7 @@
 #include <types/global.h>
 #include <types/pattern.h>
 
+#include <proto/log.h>
 #include <proto/pattern.h>
 #include <proto/sample.h>
 
@@ -1723,6 +1724,34 @@ int pat_ref_add(struct pat_ref *ref,
 	return 1;
 }
 
+/* This function prune <ref>, replace all reference by the references
+ * of <replace>, and reindex all the news values.
+ *
+ * The pattern are loaded in best effort and the errors are ignored,
+ * but writed in the logs.
+ */
+void pat_ref_reload(struct pat_ref *ref, struct pat_ref *replace)
+{
+	struct pattern_expr *expr;
+	struct pat_ref_elt *elt;
+	char *err = NULL;
+
+	pat_ref_prune(ref);
+
+	LIST_ADD(&replace->head, &ref->head);
+	LIST_DEL(&replace->head);
+
+	list_for_each_entry(elt, &ref->head, list) {
+		list_for_each_entry(expr, &ref->pat, list) {
+			if (!pat_ref_push(elt, expr, 0, &err)) {
+				send_log(NULL, LOG_NOTICE, "%s", err);
+				free(err);
+				err = NULL;
+			}
+		}
+	}
+}
+
 /* This function prune all entries of <ref>. This function
  * prune the associated pattern_expr.
  */
-- 
1.8.3.2

